### PR TITLE
Improve typings for TypeScript.

### PR DESCRIPTION
**Pull request by barinbritva**
I specified signatures for all classes and methods.
Also, I have a few questions which I will write in the comments.

### CODE REVIEW

File: lib/types/index.d.ts [hunk 1]
```diff
@@ -1,98 +1,130 @@
...
-    }
+        hasParameter (key: string): boolean;
 
-    export class Definition {
-        constructor (...args: any[]);
+        isSet (id: string): boolean;
+
+        register (id: string, object: any, args: Argument[]): Definition;
+
+        registerExtension (extension: Extension): void;
```
Comment (barinbritva) [inline lib/types/index.d.ts]: What are extensions? What do they? I didn't find any information in the documentation and couldn't understand it from the code.

File: lib/types/index.d.ts [hunk 2]
```diff
@@ -1,98 +1,130 @@
...
+        hasDefinition (key: string): boolean;
 
-    }
+        hasParameter (key: string): boolean;
 
-    export class Definition {
-        constructor (...args: any[]);
+        isSet (id: string): boolean;
+
+        register (id: string, object: any, args: Argument[]): Definition;
```
Comment (barinbritva) [inline lib/types/index.d.ts]: register method describes id as string or number in jsdoc, but in other methods id is defined as string only. Should we fix jsdoc? Or does it make a sense?

File: lib/types/index.d.ts [hunk 3]
```diff
@@ -1,98 +1,130 @@
...
+        set (id: string, instance: any): void;
 
-        addTag (name: any, ...args: any[]): any;
+        setAlias (alias: string, id: string): void;
 
-        isPublic (...args: any[]): any;
+        setDefinition (id: string, definition: Definition): Definition;
 
-        setFactory (Object: any, method: any): void;
+        setParameter (key: string, value: Parameter): void;
```
Comment (barinbritva) [inline lib/types/index.d.ts]: setParameter method describes value as string or array, but in the code, it also can be boolean and object. I specified it as: type Parameter = string|boolean|object|any[]; Maybe we should fix jsdoc too.

File: lib/types/index.d.ts [hunk 4]
```diff
@@ -1,98 +1,130 @@
...
+        logger: Logger;
 
-        isSet (id: any): any;
+        addCompilerPass (compilerPass: any, type?: PassConfigHook, priority?: number): void;
 
-        register (id: any, ...args: any[]): any;
+        compile (): void;
 
-        registerExtension (extension: any): void;
+        findDefinition (key: string): Definition;
```
Comment (barinbritva) [inline lib/types/index.d.ts]: findDefinition method describes return value as Promise, but get method of Map instance isn't asynchronous. getDefinition which also uses get method from this._definitions describes correct return value - Definition. Please correct me if I'm wrong. Maybe I miss something

File: lib/types/index.d.ts [hunk 5]
```diff
@@ -1,98 +1,130 @@
...
+        addMethodCall (method: string, args: any[]): Definition;
+
+        addProperty (key: string, value: any): Definition;
+
+        addTag (name: string, attributes: Map<any, any>): Definition;
+
+        isPublic (bypassPublic?: boolean): boolean;
 
-        load (...args: any[]): void;
+        setFactory (Object: Reference|Object, method: string): void;
```
Comment (barinbritva) [inline lib/types/index.d.ts]: Should I include PackageReference too? Or is it ok?

File: lib/types/index.d.ts (hunk 1 above)
Comment (zazoomauro) [reply lib/types/index.d.ts]: to be honest... no idea. I did that 3 years ago and I don't remember the real purpose of that feature.

File: lib/types/index.d.ts (hunk 2 above)
Comment (zazoomauro) [reply lib/types/index.d.ts]: change it. string is more correct

File: lib/types/index.d.ts (hunk 3 above)
Comment (zazoomauro) [reply lib/types/index.d.ts]: yes. is a mistake on jsdoc

File: lib/types/index.d.ts (hunk 4 above)
Comment (zazoomauro) [reply lib/types/index.d.ts]: you are right. no promise here

File: lib/types/index.d.ts (hunk 5 above)
Comment (zazoomauro) [reply lib/types/index.d.ts]: nope! PackageReference ain't go in that way.

File: lib/types/index.d.ts (hunk 1 above)
Comment (barinbritva) [reply lib/types/index.d.ts]: Ok then :)

**Pull request merged**